src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs

```diff
@@ -279,21 +279,33 @@ private async Task ExecuteAsync(HttpContext context, ConnectionDelegate connecti
         private async Task ProcessNegotiate(HttpContext context, HttpConnectionDispatcherOptions options, ConnectionLogScope logScope)
         {
             context.Response.ContentType = "application/json";
+            string error = null;
+            int clientProtocolVersion = 0;
+            if (context.Request.Query.TryGetValue("NegotiateVersion", out var queryStringVersion))
+            {
+                // Set the negotiate response to the protocol we use.
+                var queryStringVersionValue = queryStringVersion.ToString();
+                if (!int.TryParse(queryStringVersionValue, out clientProtocolVersion))
+                {
+                    error = $"The client requested an invalid protocol version '{queryStringVersionValue}'";
+                    Log.InvalidNegotiateProtocolVersion(_logger, queryStringVersionValue);
+                }
+            }
 
             // Establish the connection
-            var connection = CreateConnection(options, context);
+            var connection = CreateConnection(options, context, clientProtocolVersion, error);
```

Comment:
Nit: If the only reason for passing in an error string here is to return null when it's set, I think it would be clearer to just do the null check here.